crypto/pqt/pqt_kem.c

```diff
@@ -0,0 +1,498 @@
+                                            size_t secret_key_len,
+                                            const EC_GROUP *group) {
+  BIGNUM *secret_key_num = BN_bin2bn(secret_key, secret_key_len, NULL);
+  EC_KEY *eckey = EC_KEY_new();
+  if ((secret_key_num == NULL) || (eckey == NULL) ||
+      !EC_KEY_set_group(eckey, group) ||
+      !EC_KEY_set_private_key(eckey, secret_key_num)) {
+    BN_free(secret_key_num);
+    EC_KEY_free(eckey);
+    return NULL;
+  }
+
+  BN_free(secret_key_num);
+  return eckey;
+}
+
+// NIST-P public keys are elliptic curve points. This function does validity
+// checks on the provided |eckey|, then writes the public key to |public_key|
+// as an uncompressed point, to length |secret_key_len|.
+//
+// It matches SerializePublicKey in RFC 9180.
+static int nistp_serialize_public_key(uint8_t *public_key,
+                                      size_t public_key_len,
+                                      const EC_KEY *eckey) {
+  if (!EC_KEY_check_fips(eckey) ||
+      (EC_POINT_point2oct(EC_KEY_get0_group(eckey),
+                          EC_KEY_get0_public_key(eckey),
+                          POINT_CONVERSION_UNCOMPRESSED, public_key,
+                          public_key_len, NULL) != public_key_len)) {
+    return 0;
+  }
+  return 1;
+}
+
+// This function parses the |public_key| buffer back into an elliptic curve
+// point, checks that it is not zero, and returns a freshly allocated |EC_KEY|
+// on success, and NULL on error.
+//
+// It matches DeserializePrivateKey in RFC 9180.
+static EC_KEY *nistp_deserialize_public_key(const uint8_t *public_key,
+                                            size_t public_key_len,
+                                            const EC_GROUP *group) {
+  EC_POINT *point = EC_POINT_new(group);
+  EC_KEY *eckey = EC_KEY_new();
+  if ((point == NULL) || (eckey == NULL) ||
+      !EC_POINT_oct2point(group, point, public_key, public_key_len, NULL) ||
+      !EC_KEY_set_group(eckey, group) || !EC_KEY_set_public_key(eckey, point) ||
+      !EC_KEY_check_fips(eckey)) {
+    EC_POINT_free(point);
+    EC_KEY_free(eckey);
+    return NULL;
+  }
+
+  EC_POINT_free(point);
+  return eckey;
+}
+
+static int nistp_keygen_deterministic(const EC_GROUP *group,
+                                      uint8_t *public_key,
+                                      size_t public_key_len,
+                                      uint8_t *secret_key,
+                                      size_t secret_key_len,
+                                      const uint8_t *seed, size_t seed_len) {
+  EC_KEY *eckey = nistp_internal_keygen_deterministic(group, seed, seed_len);
+  if ((eckey == NULL) ||
+      !nistp_serialize_secret_key(secret_key, secret_key_len, eckey) ||
+      !nistp_serialize_public_key(public_key, public_key_len, eckey)) {
+    EC_KEY_free(eckey);
+    return 0;
+  }
+  EC_KEY_free(eckey);
+  return 1;
+}
+
+static int nistp_encaps_deterministic(
```

Comment:
Some of these functions does a lot of `EC_KEY_check_fips()`.

For example, this function invokes `EC_KEY_check_fips()` through
```
- nistp_internal_keygen_deterministic
- nistp_serialize_public_key
- nistp_deserialize_public_key
- ECDH_compute_key_fips
```
though not all on the same input.

Similarly, `nistp_keygen_deterministic()` invokes `EC_KEY_check_fips()` a total of 3 times on the same object.

Reply:
I was being extremely defensive when I first wrote these functions.

First, I removed the checks from `serialize_public_key` and `serialize_secret_key`, where they are totally unnecessary.

Second, I removed the check from `nistp_deserialize_public_key` and renamed it to `nistp_deserialize_public_key_unchecked` and added comments at call sites noting why unchecked is okay.

After these changes, `nistp_encaps_deterministic` calls `EC_KEY_check_fips` twice, once on the encapsulator `eckey` (inside `nistp_internal_keygen_deterministic`) and once on the `peer_eckey` (inside `ECDH_compute_key_fips`). 

And `nistp_internal_keygen_deterministic` only calls `EC_KEY_check_fips` once.